src/Stryker.Core/Stryker.Core/Mutants/MutationContext.cs

```diff
@@ -47,10 +56,28 @@ public MutationContext Clone()
             return new MutationContext(this);
         }
 
+        public void Discard()
```

Comment:
I'm not 100 percent sure the context should be responsible for deciding to discard. Especially since half the responsibility is also with the MemberDeclarationOrchestrator.

It feels both the right and wrong location if you get what I mean.

Reply:
I especially dislike that it changes the mutants statuses. I do like that it makes the decision that it's neccesary based on HasStatementLevelMutant. But then it only gets to decide to do so when MemberDeclarationOrchestrator triggers the call.

Reply:
Thanks for this remarks.
We have a complex relationship with mutators, mutants, orchestrator, injection engines and MutantContext, with somewhat overlapping responsibilities.
I don't think that is an issue in itself: what can be/should be done through source code mutation is still unclear, so it reflects in the design.
As documented elsewhere, I am not happy with this design either, but I have no idea (yet) for a better one, hence I prefer not to defer a bug fix till we are happy with a design.
. I think that deserve a specific channel (here or Slack) to discuss this matter further.

Coming back to your remarks:
> I'm not 100 percent sre the context should be responsible for deciding to discard. Especially since half the responsibility is also with the MemberDeclarationOrchestrator.

We can inject mutation as long as we can:
1. inject some control logic
2. duplicate everything that is 'below' the control logic level
3. it results on something compilable
(1) and (2) depends on available injection engine, and (3) depends (mostly) on the mutator and (somewhat) on roll back logic.
As of now, we have to inject the mutations within the very same member the original code is, hence adding a _MemberOrchestrator_ having the responsibility to clear everything. But this is not right, it means that if we find a way to have more mutations by duplicating whole members or even whole classes, we have to think to get rid of that logic.
An alternative would be to prevent injection engine to try and inject mutant where they do not belong (currently this triggers an exception).

 > I especially dislike that it changes the mutants statuses.
 I understand, but we have to do this because there is no distinction in Stryker between mutations and mutants. Mutations are promoted to full mutant status as soon as they are built. 
 I know the original design tackled this by cleverly using lazy enumeration, but I was not able to keep this design when I refactored the orchestration.
I think a better design would be to somehow promote mutation to mutants when they are inserted, not generated.

As signaled above, thanks again for these insightful questions and looking forward discussing design somewhere.

Reply:
I agree that it is not blocking for merging this but wanted to make a note regardless :)